book/src/linking_and_dividing.md

```diff
@@ -0,0 +1,337 @@
+finishing this chapter, you'll have written a method that divides two integers
+using Rust.
+
+# Whats in NativeAPI?
+
+`NativeAPI` is a class that lives in `jnibookjava`, the Java side of the example
+code that accompanies this book. It uses `System.loadLibrary` to locate the
+`jnibookrs` lib, and specifies native methods using the `native` keyword. The
+native method called `verify_link` was used earlier to ensure that the project
+was setup correctly, with environment variables.
+
+For style reasons, we always add native methods to `NativeAPI`. It doesn't need
+to be that way, it's just a convention that we follow here.
+
+```java
+package com.github.jni_rs.jnibook;
+
+class NativeAPI {
+
+   private static final Throwable INIT_ERROR;
+
+   // The static block will be executed the first time the NativeAPI
+   // class is used.
+   static {
+       Throwable error = null;
+       try {
+           System.loadLibrary("jnibookrs");
+       } catch (Throwable t) {
+           error = t;
+       }
+       INIT_ERROR = error;
+   }
+
+   private NativeAPI() {
+       // Not instantiable
+   }
+
+   static native int verify_link();
+}
+```
+
+
+## Function Linking Rules
+
+JNI identifies the functions it should link to using the package path, class
+name, and method name by following [these
+rules](https://docs.oracle.com/en/java/javase/11/docs/specs/jni/design.html).
+Since `.` isn’t allowed in C function names, the Java function's package path's
+`.` is mapped to `_`. However, since `_` could appear in a function name, it
+needs a different representation. So `_` gets replaced by `_1` before replacing
+`.` with `_`. And finally, the function name has to be prefixed with `Java_`, to
+avoid conflicts with other languages that support similar naming schemes.
+
+## LookingGlass - Function Linking Rules
+
+Since the naming rules are esoteric, you'll practice them as they are introduced
+using an example called `LookingGlass`. This example is purely for discussion
+and it won't be necessary to code it. We'll return back to the `LookingGlass`
+example for "Parameters" and "ABI and disabling mangling."
+
+The definition of `LookingGlass` on the Java side will be:
+
+```java
+package com.github.imaginarypackagename;
+
+class LookingGlass {
+
+    native static void test_call();
+}
+```
+
+And now, lets apply the naming rules discussed in "Function Linking Rules." We'll start with this code in Rust:
+
+```rust
+fn test_call() {
+
+}
+```
+
+Prefix the Rust function name with `Java.` +
+   `com.github.imaginarypackagename.` + `LookingGlass`.
+
+```rust
+fn Java.com.github.imaginarypackagename.LookingGlass.test_call() {
+
+}
+```
+
+Then, replace `_` with `_1`.
+
+```rust
+fn Java.com.github.imaginarypackagename.LookingGlass.test_1call() {
+
+}
+```
+
+Finally, replace `.` with `_`.
+
+```rust
+fn Java_com_github_imaginarypackagename_LookingGlass_test_1call() {
+
+}
+```
+
+Now the native function is named correctly for JNI.
+
+## Parameters and Return Types
```

Comment:
Would it be clearer if first a 1st argument (`JNIEnv`) is desribed; 
then — the 2nd argument:
- for native static methods (a reference to the defining class);
- for native instance methods (a reference to `this`)?

Reply:
Yes it would. The writing there is kind of odd, in that it talks about argument 2, then 1, then 2. Will fix.

Reply:
Updated this part